File: solutions/c/armstrong-numbers/2/armstrong_numbers.c
#include "armstrong_numbers.h"
#include <math.h>

bool is_armstrong_number(int candidate) {
    if (candidate < 10) {
        return true;
    }

    int digit_count = log10(candidate) + 1;

    int num = candidate;
    int result = 0;
    while (num > 0) {
        int digit = num % 10;
        int partial_result = 1;
        for (int i = 0; i < digit_count; i++)
            partial_result *= digit;
        result += partial_result;
        num /= 10;
    }
    return candidate == result;
}

File: solutions/clojure/armstrong-numbers/2/src/armstrong_numbers.clj
(ns armstrong-numbers)

(defn digits 
  ([num] (digits num []))
  ([num acc] (let [[digit rest] [(mod num 10) (-> num (/ 10.0) int)]]
               (if (= digit 0) [acc (count acc)]
                   (digits rest (conj acc digit))))))

(defn sum [[digits digits-count]] (->> digits
                                       (map #(Math/pow % digits-count))
                                       (reduce +)))

(defn armstrong? [num] (->> num
                            digits
                            sum
                            (== num)))
